### DISCUSSION TITLE
Invalid work «$watch»

### DISCUSSION THREAD

**Discussion by maxsite** [alpinejs/alpine#1001] [4. General]
Demo: https://codepen.io/maxsite/pen/OJRoOBd

Click Add, Add. In console:
`"Watch (value): [{'bla':1},{'bla':2}]"`

This right.

Click Del:
```
"Watch (value): [{'bla':1},null]"
"Real todos: [{'bla':1}]"
```

$watch value incorect: Added element null.

**Comment by SimoTod** [comment]
Good point. It's a bit due to the way splice works.
It reindexed all the allowed items first, then it sets to null whatever is left at the end of the array and eventually it adjust the length.
Each of these operations will trigger a mutation except for the last one (by design, Alpine is not triggering mutations when length changes to avoid duplicate watcher calls). you can see that behaviour if you add 3 or more items.

If you really need to use splice, you'll need a more sophisticated logic in your watcher.

**Comment by maxsite** [comment]
Thanks! 

Then instead of splice  you can use filter

`todos = todos.filter((item, key) => key !== 0);`
